sql: Introduce the concept of system cluster privileges

RichardJCai opened the pull request:
System Cluster Privileges (global) privileges are the first
type of privileges that use the system.privileges table.

Release note (sql change): Introduce SYSTEM CLUSTER PRIVILEGES.
These are "global" privileges that live above the database level.

Example: `GRANT SYSTEM MODIFYCLUSTERSETTING TO foo`

Currently `MODIFYCLUSTERSETTING` is the only system cluster privilege,
it allows users to query the `crdb_internal.cluster_settings` table.

More functionality will be gated behind having `MODIFYCLUSTERSETTING` privilege
in the future. This is not the same as the `MODIFYCLUSTERSETTING` role option.

RichardJCai committed:
sql: add new system.privileges table

First step in upcoming privileges work.
Explained in RFC: https://github.com/cockroachdb/cockroach/pull/80580

Release note: None

RichardJCai committed:
sql: support "GLOBAL" privileges

Global privileges are the first
type of privileges that use the system.privileges table.

More functionality will be gated behind having `MODIFYCLUSTERSETTING` privilege
in the future. This is not the same as the `MODIFYCLUSTERSETTING` role option.

Release note (sql change): Introduce GLOBAL privileges.
These are privileges that live above the database level.

Example: `GRANT SYSTEM MODIFYCLUSTERSETTING TO foo`

Currently `MODIFYCLUSTERSETTING` is the only global privilege,
it allows users to query the `crdb_internal.cluster_settings` table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82166)
<!-- Reviewable:end -->

RichardJCai commented:
Need to add some more tests / a bit more cleanup but I think this is ready for a first pass of reviews right now.

RichardJCai commented on pkg/sql/crdb_internal.go:
```diff
@@ -1556,9 +1556,13 @@ CREATE TABLE crdb_internal.cluster_settings (
+				// role option OR the MODIFYCLUSTERSETTING system cluster privilege.
+				// We return the error for "system cluster privilege" due to
+				// the long term goal of moving away from coarse-grained role options.
+				if err := p.CheckPrivilege(ctx, &catalog.SystemClusterPrivilege{}, privilege.MODIFYCLUSTERSETTING); err != nil {
```
Reminder to myself, this needs to be version gated.

ecwall reviewed (changes requested):
> Reviewed 40 of 49 files at r1, 8 of 50 files at r3, 45 of 45 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @msbutler, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke.go` line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N43yCt18JaKAmiCu_yQ:-N43yCt18JaKAmiCu_yR:b-eo5fhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/grant_revoke.go#L154)):*
> ```Go
> }
> 
> type changeNonDescriptorBackedPrivilegesNode struct {
> ```

Can you use composition here with a struct that contains the shared fields between this and `changePrivilegesNode` and embed it in both?

`preChangePrivilegesValidation` could be a method with `changePrivilegesNode` as a receiver. 
___
*[`pkg/sql/grant_revoke_system.go` line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N4426mJ1x8S2hejgnFS:-N4426mJ1x8S2hejgnFT:bxekxlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/grant_revoke_system.go#L118)):*
> ```Go
> 			systemPrivilegeObject.ToString(),
> 			bitarray.MakeBitArrayFromInt64(64, int64(userPrivs.Privileges), 64),
> 			bitarray.MakeZeroBitArray(64),
> ```

Are grant options part of the table but unused? (same question for grant above)
___
*[`pkg/sql/catalog/catprivilege/validate.go` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N441_GhB0gPZkH8Ds8s:-N441_GhB0gPZkH8Ds8t:b-9xigmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/catalog/catprivilege/validate.go#L68)):*
> ```Go
> 	}
> 
> 	if !reflect.DeepEqual(out, systemPrivilegeObject) {
> ```

Does `!=` not work here?
___
*[`pkg/sql/privilege/privilege.go` line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N4437iKEzbPy05rMxd8:-N4437iL0vXuyQuVlNcS:b42pihi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/privilege/privilege.go#L46)):*
> ```Go
> 	CONNECT              Kind = 11
> 	RULE                 Kind = 12
> 	MODIFYCLUSTERSETTING Kind = 13 // SYSTEM CLUSTER PRIVILEGE.
> ```

Is there a concern about adding non-standard privilege kinds here? It looks like we are limited to 32 possible values `type Kind uint32`.

I haven't thought about this much, but is it possible to reuse the `UPDATE` privilege and add it to a "cluster setting" (does something like this exist?) database object similar to table, schema, etc?
___
*[`pkg/util/bitarray/bitarray.go` line 622 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N447Px07_-hpfROrUf4:-N447Px07_-hpfROrUf5:b-xe1t1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/util/bitarray/bitarray.go#L622)):*
> ```Go
> // AsUInt64 returns the uint64 constituted from the rightmost bits in the
> // bit array.
> func (d BitArray) AsUInt64() uint64 {
> ```

Does having a value receiver end up copying the array?


<!-- Sent from Reviewable.io -->

knz commented:
> More functionality will be gated behind having MODIFYCLUSTERSETTING privilege
> in the future. This is not the same as the MODIFYCLUSTERSETTING role option.

Sorry I missed the memo; what's the difference between the two? (Also I'm pretty sure that difference needs to be explained in a release note.)

RichardJCai commented:
> > More functionality will be gated behind having MODIFYCLUSTERSETTING privilege
> > in the future. This is not the same as the MODIFYCLUSTERSETTING role option.
> 
> Sorry I missed the memo; what's the difference between the two? (Also I'm pretty sure that difference needs to be explained in a release note.)

Functionally they're going to be the same thing, the goal is to emphasize using the privilege and not the role option and in general move away from and eventually deprecate these non granular role options (the ones we added somewhat adhoc and are not in pg)

knz commented:
> Functionally they're going to be the same thing

if that part of the sentence is true, then the sentence "More functionality will be gated behind ..." should be moved from the release note back up higher in the commit message, outside of the release note. Because that means that change is not user-facing.

RichardJCai reviewed (commented):
Updated the commit msg.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-:-N4cdIeOEz2kUYWh8DN3:b-vuny9v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, and @rafiss)

___
*[`pkg/sql/grant_revoke.go` line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N43yCt18JaKAmiCu_yQ:-N4cb-zD8Af1RPxIU7DT:b-lhdfkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/grant_revoke.go#L154)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Can you use composition here with a struct that contains the shared fields between this and `changePrivilegesNode` and embed it in both?

`preChangePrivilegesValidation` could be a method with `changePrivilegesNode` as a receiver. 
</blockquote></details>

Sounds reasonable, done.
___
*[`pkg/sql/grant_revoke_system.go` line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N4426mJ1x8S2hejgnFS:-N4cb2nu4w9TwCX4hH5j:b-wrrs98) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/grant_revoke_system.go#L118)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Are grant options part of the table but unused? (same question for grant above)
</blockquote></details>

I'm gonna add a todo to add grant options in a follow up, don't wan tthis PR to get too big.
___
*[`pkg/sql/catalog/catprivilege/validate.go` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N441_GhB0gPZkH8Ds8s:-N4cbg990Thr9YbuXTN7:b-qdefth) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/catalog/catprivilege/validate.go#L68)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Does `!=` not work here?
</blockquote></details>

`systemPrivilegeObject catalog.SystemPrivilegeObject` is an interface here so `!=` for the nil case can be weird I think, https://yourbasic.org/golang/interfaces-explained/ see the Equality part.

I think `DeepEqual` safer to use.
___
*[`pkg/sql/privilege/privilege.go` line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N4437iKEzbPy05rMxd8:-N4ccTz7EJCVhnkpOl4a:b8rfb89) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/sql/privilege/privilege.go#L46)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Is there a concern about adding non-standard privilege kinds here? It looks like we are limited to 32 possible values `type Kind uint32`.

I haven't thought about this much, but is it possible to reuse the `UPDATE` privilege and add it to a "cluster setting" (does something like this exist?) database object similar to table, schema, etc?
</blockquote></details>

This is a valid concern, although I think it should be fairly easy to go from uint32 to uint64 or uint128 even though we store the ints in the protobuf. Updating a proto field from uint32 is uint64 is straightforward I think cause it's wire compatible. If we want to avoid this we can separate out system privileges here but it would be slightly annoying from a code standpoint.
___
*[`pkg/util/bitarray/bitarray.go` line 622 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N447Px07_-hpfROrUf4:-N4cdGRVBYTK_lcS4DJw:b-fpmdgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d96b81ab9d03ca97207c9479495e6307bfc08404/pkg/util/bitarray/bitarray.go#L622)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Does having a value receiver end up copying the array?
</blockquote></details>

Yeah,going to pass a ptr.


<!-- Sent from Reviewable.io -->

RichardJCai commented:
Gonna make a follow up issue to support `GRANT OPTIONS` as well as adding other privileges. Trying to keep this PR relatively contained.

rafiss reviewed (changes requested):
> Dismissed @RichardJCai from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, and @RichardJCai)

___
*[`pkg/sql/authorization.go` line 194 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55p9JO7iQrOcsYjzqy:-N55p9JPBEAYGmmDUcdl:bmbjxp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/authorization.go#L194)):*
> ```Go
> 
> 	hasPriv, err := p.checkRolePredicate(ctx, user, func(role username.SQLUsername) bool {
> 		return privs.CheckGrantOptions(role, privList) || (descriptor != nil && IsOwner(descriptor, role))
> ```

nit: maybe a comment that says when descriptor is nil would help
___
*[`pkg/sql/authorization.go` line 236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55pzp29EyrHVzjVIc7:-N55pzp29EyrHVzjVIc8:b-u4lall) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/authorization.go#L236)):*
> ```Go
> }
> 
> // IsOwner returns if the role has ownership on the descriptor.
> ```

nit: update the comment to change "descriptor" -> "privilege object" and mention that admin is always the owner for system privileges

also, could update this function so the nil check is done here instead of at the call site
___
*[`pkg/sql/crdb_internal.go` line 1564 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55qUBX3GNJsXzvU3JN:-N55qUBX3GNJsXzvU3JO:bm8117d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/crdb_internal.go#L1564)):*
> ```Go
> 					// We return the error for "system cluster privilege" due to
> 					// the long term goal of moving away from coarse-grained role options.
> 					if err := p.CheckPrivilege(ctx, &catalog.SystemClusterPrivilege{}, privilege.MODIFYCLUSTERSETTING); err != nil {
> ```

not following this; why use an empty struct?
___
*[`pkg/sql/grant_revoke.go` line 175 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55rJOZDLbJsV8RsY-K:-N55rJOZDLbJsV8RsY-L:b-hcwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke.go#L175)):*
> ```Go
> 
> func (n *changePrivilegesNode) preChangePrivilegesValidation(
> 	params runParams, isGrant bool, withGrantOption bool, grantees []username.SQLUsername,
> ```

nit: i don't think isGrant, withGrantOption, and grantees need to be params. aren't they all available from the receiver?
___
*[`pkg/sql/grant_revoke_system.go` line 33 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55s3ZlEpjVBXvHpaXY:-N55s3ZlEpjVBXvHpaXZ:b-l8l68) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L33)):*
> ```Go
> func (n *changeNonDescriptorBackedPrivilegesNode) ReadingOwnWrites() {}
> 
> func (n *changeNonDescriptorBackedPrivilegesNode) startExec(params runParams) error {
> ```

i think we'd want a version gate in here (or somewhere along this code path)
___
*[`pkg/sql/grant_revoke_system.go` line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55sHBtCHRA60E6bIhu:-N55sHBtCHRA60E6bIhv:bn3h54d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L60)):*
> ```Go
> 			userPrivs, found := syntheticPrivDesc.FindUser(user)
> 			if !found {
> 				return errors.AssertionFailedf("user %s not found", user)
> ```

nit: use pgerror here?
___
*[`pkg/sql/grant_revoke_system.go` line 71 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55sSME8NF2xmkC-q_F:-N55sSME8NF2xmkC-q_G:b-sxalh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L71)):*
> ```Go
> 				user.Normalized(),
> 				systemPrivilegeObject.ToString(),
> 				bitarray.MakeBitArrayFromInt64(64, int64(userPrivs.Privileges), 64),
> ```

nit: should it use uint64?
___
*[`pkg/sql/grant_revoke_system.go` line 117 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55siwfAHAgfh6zceyl:-N55siwh1z2ELwfl7MvT:bm5pjjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L117)):*
> ```Go
> 			user.Normalized(),
> 			systemPrivilegeObject.ToString(),
> 			bitarray.MakeBitArrayFromInt64(64, int64(userPrivs.Privileges), 64),
> ```

just wondering what happens under the hood. this is a VARBIT right? so why does it need a fixed width here?

also, seems a bit odd that we have to make it a signed int, but i guess that's how the function works?
___
*[`pkg/sql/catalog/privilege_object.go` line 24 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55tGBg8hTBzA45LAZu:-N55tGBg8hTBzA45LAZv:b3dpz42) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/catalog/privilege_object.go#L24)):*
> ```Go
> type PrivilegeObject interface {
> 	GetPrivilegeDescriptor(ctx context.Context, planner eval.Planner) (*catpb.PrivilegeDescriptor, error)
> 	GetObjectType() string
> ```

should this return an enum or does it not matter that much?
___
*[`pkg/sql/catalog/privilege_object.go` line 25 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55tARs1brUFZey-V-1:-N55tARs1brUFZey-V-2:b-48iom7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/catalog/privilege_object.go#L25)):*
> ```Go
> 	GetPrivilegeDescriptor(ctx context.Context, planner eval.Planner) (*catpb.PrivilegeDescriptor, error)
> 	GetObjectType() string
> 	GetName() string
> ```

nit: function comments would be useful here i think
___
*[`pkg/sql/catalog/systemschema/system.go` line 2412 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55v1th9pEx8Gsqa5k5:-N55v1th9pEx8Gsqa5k6:b-an5rl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/642539b04c3a039034f8f35f03fe0fec77b3a602/pkg/sql/catalog/systemschema/system.go#L2412)):*
> ```Go
> 	)
> 
> 	SystemPrivilegeTable = registerSystemTable(
> ```

i believe there's something where we tell backup/restore how to handle this system table (opt in or out). just checking if we need that
___
*[`pkg/sql/privilege/privilege.go` line 118 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55u6v599qU3F4EI4zj:-N55u6v6Dugp2_oUb7hj:bljs2oo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/privilege/privilege.go#L118)):*
> ```Go
> }
> 
> // ByValue is just an array of privilege kinds sorted by value.
> ```

hm, it says sorted by value, but doesn't seem to be true. should we adjust the comment?
___
*[`pkg/sql/logictest/testdata/logic_test/system_privileges` line 21 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55uTiL4xoA0r5PGWoW:-N55uTiL4xoA0r5PGWoX:b-7rib2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/logictest/testdata/logic_test/system_privileges#L21)):*
> ```
> SELECT * FROM system.privileges
> ----
> testuser  /system/  0000000000000000000000000000000000000000000000000010000000000000  0000000000000000000000000000000000000000000000000000000000000000
> ```

didn't think about this in the first commit: but should we make the system table store something more readable, like an array of text? i'm not sure the space savings make this output worth it
___
*[`pkg/sql/logictest/testdata/logic_test/system_privileges_mixed` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55unoe126p3Y7gFfwl:-N55unoe126p3Y7gFfwm:b-ak3vbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/logictest/testdata/logic_test/system_privileges_mixed#L1)):*
> ```
> # LogicTest: local-mixed-22.1-22.2
> ```

should make sure we can't grant system privileges in a mixed version cluster


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, and @rafiss)

___
*[`pkg/sql/authorization.go` line 194 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55p9JO7iQrOcsYjzqy:-N56dvj45lRyoZvI0p_A:bqk0pf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/authorization.go#L194)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: maybe a comment that says when descriptor is nil would help
</blockquote></details>

Updated so we don't need nil
___
*[`pkg/sql/authorization.go` line 236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55pzp29EyrHVzjVIc7:-N56ejxtAJjRiweSgL0g:b-fn7edp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/authorization.go#L236)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: update the comment to change "descriptor" -> "privilege object" and mention that admin is always the owner for system privileges

also, could update this function so the nil check is done here instead of at the call site
</blockquote></details>

Actually I just changed the method to take a `PrivilegeObject` into the receivers for `CheckGrantOptionForUser` so we don't need to check for nil anymore.

Updated comment
___
*[`pkg/sql/crdb_internal.go` line 1564 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55qUBX3GNJsXzvU3JN:-N56exauAXp_JsUfkWJO:b90xfgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/crdb_internal.go#L1564)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

not following this; why use an empty struct?
</blockquote></details>

`SystemClusterObject` doesn't have any arguments for the `catalog.PrivilegeObject` struct but future objects will have arguments. The empty struct is enough to identify the privilege object for `SystemClusterPrivilege`, for future objects we'll need more.
___
*[`pkg/sql/grant_revoke.go` line 175 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55rJOZDLbJsV8RsY-K:-N56faPe1fRB2vxMsUB0:b-fpoe3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke.go#L175)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i don't think isGrant, withGrantOption, and grantees need to be params. aren't they all available from the receiver?
</blockquote></details>

Yep you're very right



_Code quote:_
```Go
 preChangePrivilegesValidation
```
___
*[`pkg/sql/grant_revoke_system.go` line 33 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55s3ZlEpjVBXvHpaXY:-N56fgcs6o7axlMNZNk2:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L33)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we'd want a version gate in here (or somewhere along this code path)
</blockquote></details>

Done
___
*[`pkg/sql/grant_revoke_system.go` line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55sHBtCHRA60E6bIhu:-N56h4TL2KKLj95D1cxq:b-bpemaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L60)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: use pgerror here?
</blockquote></details>

I chose assertion error here since there should 100% be a user here.

```
			syntheticPrivDesc.Grant(user, n.desiredprivs, false)

			userPrivs, found := syntheticPrivDesc.FindUser(user)
			if !found {
				return errors.AssertionFailedf("user %s not found", user)
			}
```

The `Grant` above would create a user if there wasn't one.
___
*[`pkg/sql/grant_revoke_system.go` line 71 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55sSME8NF2xmkC-q_F:-N56hGW--x-vg-ojdol1:b-1vop97) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L71)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: should it use uint64?
</blockquote></details>

N/A for now after change to using string array
___
*[`pkg/sql/grant_revoke_system.go` line 117 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55siwfAHAgfh6zceyl:-N56hLSu7Fb78431MCy-:b-fc64ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L117)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just wondering what happens under the hood. this is a VARBIT right? so why does it need a fixed width here?

also, seems a bit odd that we have to make it a signed int, but i guess that's how the function works?
</blockquote></details>

Updated to str array
___
*[`pkg/sql/catalog/privilege_object.go` line 24 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55tGBg8hTBzA45LAZu:-N56h_SF4S-6PVp6uhUc:bslnblq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/catalog/privilege_object.go#L24)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this return an enum or does it not matter that much?
</blockquote></details>

We return string for `DescriptorType` which is used for this, I think string is ok.

Or do you mean create constants for this somewhere?

```
// DescriptorType is a symbol representing the (sub)type of a descriptor.
type DescriptorType string

const (
	// Any represents any descriptor.
	Any DescriptorType = "any"

	// Database is for database descriptors.
	Database = "database"

	// Table is for table descriptors.
	Table = "relation"

	// Type is for type descriptors.
	Type = "type"

	// Schema is for schema descriptors.
	Schema = "schema"
)
```
___
*[`pkg/sql/catalog/systemschema/system.go` line 2412 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55v1th9pEx8Gsqa5k5:-N56im906uSAC8a2nAsI:b-kibn2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/642539b04c3a039034f8f35f03fe0fec77b3a602/pkg/sql/catalog/systemschema/system.go#L2412)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i believe there's something where we tell backup/restore how to handle this system table (opt in or out). just checking if we need that
</blockquote></details>

Yeah this is defined in `system_schema.go`

right now we're opting out (consistent with regular privileges) but we probably want to add logic to optionally include this in backup and restore.

```
	systemschema.SystemPrivilegeTable.GetName(): {
		shouldIncludeInClusterBackup: optOutOfClusterBackup,
	},
```
___
*[`pkg/sql/privilege/privilege.go` line 118 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55u6v599qU3F4EI4zj:-N56j4A14OSZZlbRQRjh:b-sr6nt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/privilege/privilege.go#L118)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, it says sorted by value, but doesn't seem to be true. should we adjust the comment?
</blockquote></details>

This is still stored by value (the INT value)
___
*[`pkg/sql/logictest/testdata/logic_test/system_privileges` line 21 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55uTiL4xoA0r5PGWoW:-N56j760DartWqdHc7RM:b-4mkj9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/logictest/testdata/logic_test/system_privileges#L21)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

didn't think about this in the first commit: but should we make the system table store something more readable, like an array of text? i'm not sure the space savings make this output worth it
</blockquote></details>

Yeah updated to string array, I don't think the VARBIT logic is worth it.
___
*[`pkg/sql/logictest/testdata/logic_test/system_privileges_mixed` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55unoe126p3Y7gFfwl:-N56jyvB49BfmiJrNkZj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/logictest/testdata/logic_test/system_privileges_mixed#L1)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should make sure we can't grant system privileges in a mixed version cluster
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, @rafiss, and @RichardJCai)

___
*[`pkg/sql/crdb_internal.go` line 1564 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55qUBX3GNJsXzvU3JN:-N5GuSXy0w-0YTIj4aWt:bf3d33) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/crdb_internal.go#L1564)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

`SystemClusterObject` doesn't have any arguments for the `catalog.PrivilegeObject` struct but future objects will have arguments. The empty struct is enough to identify the privilege object for `SystemClusterPrivilege`, for future objects we'll need more.
</blockquote></details>

would it make sense to use a singleton var for this empty `catalog.SystemClusterPrivilege` -- then we can name it in and comment it to include the explanation you just gave.
___
*[`pkg/sql/grant_revoke_system.go` line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55sHBtCHRA60E6bIhu:-N5GuvWb5RDGZKFt9kOG:bq0ncok) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/grant_revoke_system.go#L60)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I chose assertion error here since there should 100% be a user here.

```
			syntheticPrivDesc.Grant(user, n.desiredprivs, false)

			userPrivs, found := syntheticPrivDesc.FindUser(user)
			if !found {
				return errors.AssertionFailedf("user %s not found", user)
			}
```

The `Grant` above would create a user if there wasn't one.

</blockquote></details>

sounds good
___
*[`pkg/sql/catalog/privilege_object.go` line 24 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55tGBg8hTBzA45LAZu:-N5GvW0y2Xjqvb5Zvh8e:bgukvjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/catalog/privilege_object.go#L24)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

We return string for `DescriptorType` which is used for this, I think string is ok.

Or do you mean create constants for this somewhere?

```
// DescriptorType is a symbol representing the (sub)type of a descriptor.
type DescriptorType string

const (
	// Any represents any descriptor.
	Any DescriptorType = "any"

	// Database is for database descriptors.
	Database = "database"

	// Table is for table descriptors.
	Table = "relation"

	// Type is for type descriptors.
	Type = "type"

	// Schema is for schema descriptors.
	Schema = "schema"
)
```
</blockquote></details>

yeah constants could make sense then, i think
___
*[`pkg/sql/catalog/privilege_object.go` line 22 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5H13Rf8LEnKZEpmbI-:-N5H13Rf8LEnKZEpmbI0:b7xri16) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41a1b20b95bd9b4086a1eb38f5b1f373cdb25e0/pkg/sql/catalog/privilege_object.go#L22)):*
> ```Go
> // PrivilegeObject represents an object that can have privileges. The privileges
> // can either live on the descriptor or in the system.privileges table.
> type PrivilegeObject interface {
> ```

hm could you walk me though again why we need this interface? why can't we just call the "synthesize system privilege descriptor" function before the point where we need to start passing around these new PrivilegeObjects?
___
*[`pkg/sql/catalog/systemschema/system.go` line 687 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5H1KHTCcqNU2Ma6JTa:-N5H1KHTCcqNU2Ma6JTb:b-o837cl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41a1b20b95bd9b4086a1eb38f5b1f373cdb25e0/pkg/sql/catalog/systemschema/system.go#L687)):*
> ```Go
> 	path STRING NOT NULL,
> 	privileges STRING[] NOT NULL,
> 	grant_options STRING[] NOT NULL,
> ```

can the descriptor validation logic we already have be updated to make sure that there are only valid privlieges in this table (and validate each time we read from it)

ideally we could use an enum or CHECK constraint, but that would make adding a new privilege way too much of a pain i think, since each one would need a schema migration on a system table.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 1564 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55qUBX3GNJsXzvU3JN:-N5HC358EMB4Z8xxp033:b-gujf1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/crdb_internal.go#L1564)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would it make sense to use a singleton var for this empty `catalog.SystemClusterPrivilege` -- then we can name it in and comment it to include the explanation you just gave.
</blockquote></details>

DOne, created `catalog.SystemClusterPrivilegeObject`
___
*[`pkg/sql/catalog/privilege_object.go` line 24 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N55tGBg8hTBzA45LAZu:-N5HDh73ECuOXmw2iWL1:b-mp7840) ([raw file](https://github.com/cockroachdb/cockroach/blob/9db37d88b9c214ee563792c7dfc77e6c82f67cd2/pkg/sql/catalog/privilege_object.go#L24)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

yeah constants could make sense then, i think
</blockquote></details>

Added a constant for system.
___
*[`pkg/sql/catalog/privilege_object.go` line 22 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5H13Rf8LEnKZEpmbI-:-N5HCa8o2sP5pE3Gao7d:b-nee5os) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41a1b20b95bd9b4086a1eb38f5b1f373cdb25e0/pkg/sql/catalog/privilege_object.go#L22)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm could you walk me though again why we need this interface? why can't we just call the "synthesize system privilege descriptor" function before the point where we need to start passing around these new PrivilegeObjects?
</blockquote></details>

I think it simplifies a lot of with the `GetObjectType` for privilege validation.
It also simplifies some of the owner code.

This is mostly so we can identify objects and unify the code paths for descriptor backed and non-descriptor backed objects. I don't think it adds much overhead here.
___
*[`pkg/sql/catalog/systemschema/system.go` line 687 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5H1KHTCcqNU2Ma6JTa:-N5HH2_u45XAX60Fg8g_:b-8tlrd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41a1b20b95bd9b4086a1eb38f5b1f373cdb25e0/pkg/sql/catalog/systemschema/system.go#L687)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can the descriptor validation logic we already have be updated to make sure that there are only valid privlieges in this table (and validate each time we read from it)

ideally we could use an enum or CHECK constraint, but that would make adding a new privilege way too much of a pain i think, since each one would need a schema migration on a system table.
</blockquote></details>

Added a validate into the `synthesizePrivilegeDescriptorFromSystemPrivilegesTable` method


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
as discussed offline: let's have `catalog.Descriptor` implement `catalog.PrivilegeObject` so that it's easier to use as a parameter for existing functions


also hopefully minor nits, then lgtm

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-:-N5Lk2Te2gDKZIrv9vpL:bl31pxt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, @rafiss, and @RichardJCai)

___
*[`pkg/sql/catalog/system_privilege.go` line 36 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5LiSE3BPjoI55O4XC8:-N5LiSE59v7M561rJfTN:b-kpowlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5df8cbda7d5c35ce52ae424202f1ad2832b79e28/pkg/sql/catalog/system_privilege.go#L36)):*
> ```Go
> 
> // SystemClusterPrivilege represents a SystemClusterPrivilege.
> type SystemClusterPrivilege struct {
> ```

nit: should this use a type assertion instead of embedding?
```
type SystemClusterPrivilege struct {}

var _ SystemPrivilegeObject = &SystemClusterPrivilege{}
```

(actually, i guess we wouldn't even need this because of the singleton defined below)

also i'm finding it hard to keep track of the names "SystemPrivilegeObject" versus "SystemClusterPrivilege". is there a way i should be thinking about this? why do we need both?
___
*[`pkg/sql/privilege/privilege.go` line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5Ljq9f3Blw434ZzOzp:-N5Ljq9gAUcWNFhB9aJK:bc3g64u) ([raw file](https://github.com/cockroachdb/cockroach/blob/5df8cbda7d5c35ce52ae424202f1ad2832b79e28/pkg/sql/privilege/privilege.go#L48)):*
> ```Go
> 	CONNECT              Kind = 11
> 	RULE                 Kind = 12
> 	MODIFYCLUSTERSETTING Kind = 13 // SYSTEM CLUSTER PRIVILEGE.
> ```

is this comment correct? i believe the comment will be used for the string representation


<!-- Sent from Reviewable.io -->

RichardJCai committed:
sql: update HasOwnership calls to use PrivilegeObject

misc lint changes and updating TestState to use the correct
method signature for CheckPrivilege and HasOwnership.

Splitting this into a separate commit due to be able to
track bazel build succeeding despite TestState not
implementing AuthorizationAccessor correctly.

Release note: None

RichardJCai reviewed (commented):
>as discussed offline: let's have catalog.Descriptor implement catalog.PrivilegeObject so that it's easier to use as a parameter for existing functions

This was the case but forgot to update `TestState`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-:-N5M8gCy6M--Qy0jLn60:b884nnf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, and @rafiss)

___
*[`pkg/sql/catalog/system_privilege.go` line 36 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5LiSE3BPjoI55O4XC8:-N5Lxs357Ff8ECGXXa8M:b-74tjsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/5df8cbda7d5c35ce52ae424202f1ad2832b79e28/pkg/sql/catalog/system_privilege.go#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: should this use a type assertion instead of embedding?
```
type SystemClusterPrivilege struct {}

var _ SystemPrivilegeObject = &SystemClusterPrivilege{}
```

(actually, i guess we wouldn't even need this because of the singleton defined below)

also i'm finding it hard to keep track of the names "SystemPrivilegeObject" versus "SystemClusterPrivilege". is there a way i should be thinking about this? why do we need both?
</blockquote></details>

The mental model is `SystemPrivilegeObject` is a privilege object where the privileges are stored in the system table.

`SystemClusterPrivilege` is a type of `SystemPrivilegeObject`, there will be more types in the future (jobs, etc). 

I can see that the naming is definitely confusing though, `SystemClusterPrivilege` is a "global" privilege.

Let me know if you can think of better naming solutions for these.

Also I removed the embedding since we do already have the type assertion.
___
*[`pkg/sql/privilege/privilege.go` line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5Ljq9f3Blw434ZzOzp:-N5LyXaO6Ly72WM8VtVZ:b-jytxep) ([raw file](https://github.com/cockroachdb/cockroach/blob/5df8cbda7d5c35ce52ae424202f1ad2832b79e28/pkg/sql/privilege/privilege.go#L48)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this comment correct? i believe the comment will be used for the string representation
</blockquote></details>

I checked `kind_string.go` and that didn't seem to be the case but I'm happy to just remove this comment, it isn't adding anything.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, and @rafiss)

___
*[`pkg/sql/privilege/privilege.go` line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5Ljq9f3Blw434ZzOzp:-N5MH89Y7OlmFBqCu-Xe:bnbf273) ([raw file](https://github.com/cockroachdb/cockroach/blob/5df8cbda7d5c35ce52ae424202f1ad2832b79e28/pkg/sql/privilege/privilege.go#L48)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I checked `kind_string.go` and that didn't seem to be the case but I'm happy to just remove this comment, it isn't adding anything.
</blockquote></details>

that's a bug because the comment from #81310 was not completely addressed

https://reviewable.io/reviews/cockroachdb/cockroach/81310#-N3R9piu3KukUO8hHuJr:-N3R9piu3KukUO8hHuJs:bd9h7vx

would you mind adding `linecomment` to `//go:generate stringer -type=Kind` above? cc @ecwall 


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, @msbutler, and @rafiss)

___
*[`pkg/sql/privilege/privilege.go` line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/82166#-N5Ljq9f3Blw434ZzOzp:-N5MOztk4CW-ISMD8RLh:b-rqra98) ([raw file](https://github.com/cockroachdb/cockroach/blob/5df8cbda7d5c35ce52ae424202f1ad2832b79e28/pkg/sql/privilege/privilege.go#L48)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

that's a bug because the comment from #81310 was not completely addressed

https://reviewable.io/reviews/cockroachdb/cockroach/81310#-N3R9piu3KukUO8hHuJr:-N3R9piu3KukUO8hHuJs:bd9h7vx

would you mind adding `linecomment` to `//go:generate stringer -type=Kind` above? cc @ecwall 
</blockquote></details>

Added it and updated the generated file.


<!-- Sent from Reviewable.io -->

RichardJCai committed:
sql: update Privilege stringer with linecomment flag

This makes DEPRECATEDGRANT resolve to just "GRANT" as a string.

Release note: None

RichardJCai committed:
sql: rename SystemPrivileges -> SyntheticPrivileges and rename SystemClusterPrivileges

Change internal representation of names.

Release note: None

RichardJCai commented:
Failure is flake, thanks for reviewing!

bors r+

RichardJCai commented:
bors r+
